File: apps/browser/extension/memory.mjs
// Augmentor — dsh-augmentor plugin, pipe, and Chromium extension
// Copyright © 2026 Manolo Remiddi
// SPDX-License-Identifier: LicenseRef-Augmentor-MIT-Resale-1.0
// License: MIT with Augmentor Resale Restriction — see LICENSE at the repository root.
import {presentSettingsForm} from './settings-form.mjs'
export function memoryDialog(doc,send,provenance,container){
  const dialog=doc.createElement('dialog');dialog.className='shared-prompt-editor memory-dialog'
  const make=(tag,text)=>{const e=doc.createElement(tag);if(text)e.textContent=text;return e}
  const intro=make('p','Relationship and work memory builds automatically from conversations. Hindsight maintains relationship pages and searchable project memory.')
  const automatic=make('fieldset'),autoNote=make('p','Loading automatic memory…'),autoContext=make('pre');autoContext.style.whiteSpace='pre-wrap';automatic.append(make('legend','Automatic continuity'),make('p','Transcripts are stored locally. Hindsight processes relationship and project context automatically, with durable retry. Voice brings relationship continuity into the conversation.'),autoNote,autoContext)
  const connection=make('fieldset'),legend=make('legend','Optional manual library connection'),fields={};connection.append(legend)
  for(const [id,label,placeholder] of [['endpoint','Endpoint','http://127.0.0.1:8888'],['apiKey','API key','Required for a remote service'],['userBank','User bank',''],['projectBank','Project bank (optional)','']]){
    const l=make('label',label),e=make('input');e.setAttribute('aria-label',label);e.placeholder=placeholder;e.type=id==='apiKey'?'password':'text';l.append(e);connection.append(l);fields[id]=e
  }
  const scope=make('select');scope.setAttribute('aria-label','Agent recall scope')
  const dataScope=make('select');dataScope.setAttribute('aria-label','Memory data scope')
  for(const select of [scope,dataScope])for(const id of ['user','project']){const option=make('option',id==='user'?'User':'Project');option.value=id;select.append(option)}
  const scopeLabel=make('label','Agent recall scope');scopeLabel.append(scope);connection.append(scopeLabel)
  connection.append(make('p','Requires Hindsight 0.9.2. Checking sends no chat history. Save enables recall. Retain sends only the text you enter below; Hindsight may charge for processing. The key is stored privately without encryption.'))
  const connectionActions=make('div'),data=make('fieldset');connection.append(connectionActions);data.append(make('legend','Memories'),dataScope)
  const text=make('textarea');text.rows=3;text.setAttribute('aria-label','Memory text');text.placeholder='Text to remember. Only this text will be retained.';data.append(text)
  const dataActions=make('div'),operationNote=make('p'),list=make('select'),contents=make('textarea'),pageNote=make('p');list.size=4;list.setAttribute('aria-label','Retained documents');contents.rows=4;contents.readOnly=true;contents.setAttribute('aria-label','Retained document text')
  data.append(dataActions,operationNote,list,contents,pageNote)
  const note=make('p','Loading memory settings…'),actions=make('div')
  dialog.append(make('h3','Memory'),intro,automatic,connection,data,note,actions);doc.body.append(dialog)
  let config={},token=null,busy=false,closed=false,rows=[],offset=0,total=0
  const button=(parent,label,fn)=>{const b=make('button',label);b.type='button';b.onclick=fn;parent.append(b);return b}
  const request=async(action,params={})=>{const r=await send('memory',{request:{action,...params}});if(!r.ok)throw Error(r.error);return r.result}
  let autoEnabled=true
  const loadAutomatic=async()=>{
    const value=await request('dual.describe');autoEnabled=!!value.enabled;autoToggle.textContent=autoEnabled?'Pause automatic memory':'Resume automatic memory';autoNote.textContent=(autoEnabled?'Automatic memory is on. ':'Automatic memory is paused. ')+`${value.events} transcript records; ${value.pending} memory scopes pending. ${value.message||''}`
    const source=provenance();if(source.sessionId){try{const context=await request('dual.recall',{session:source.harness+':'+source.sessionId});autoContext.textContent=['relationship','work'].map(kind=>kind+': '+(context[kind]?.summary||'No distilled picture yet.')).join('\n\n')}catch{autoContext.textContent='Remembered context appears after this conversation uses automatic memory.'}}
  }
  const autoToggle=button(automatic,'Pause automatic memory',async()=>{autoToggle.disabled=true;try{await request('dual.configure',{enabled:!autoEnabled});await loadAutomatic()}catch(e){autoNote.textContent=e.message}finally{autoToggle.disabled=false}})
  button(automatic,'Refresh remembered context',()=>{void loadAutomatic().catch(e=>autoNote.textContent=e.message)})
  void loadAutomatic().catch(()=>{autoNote.textContent='Automatic memory requires the updated Augmentor companion.'})
  const controls=()=>{connection.disabled=busy;data.disabled=busy;save.disabled=busy||!token;retain.disabled=busy||!config.enabled}
  const run=async(fn)=>{if(busy)return;busy=true;controls();try{await fn()}catch(e){note.textContent=e.message}finally{busy=false;controls()}}
  const invalidate=()=>{token=null;controls()};for(const e of [...Object.values(fields),scope])e.addEventListener('input',invalidate)
  const loadConfig=async()=>{
    config=await request('describe');token=null
    for(const id of ['endpoint','projectBank'])fields[id].value=config[id]??''
    fields.userBank.value=config.userBank??'augmentor-user-'+crypto.randomUUID().slice(0,12)
    fields.apiKey.value='';fields.apiKey.placeholder=config.apiKeySet?'Stored key kept if left blank':'Required for a remote service'
    scope.value=config.activeScope??'user';dataScope.value=scope.value
    note.textContent=config.enabled?'Hindsight memory enabled.':'Hindsight memory disabled.'
  }
  const operations=async()=>{
    const value=await request('operations',{scope:dataScope.value})
    const pending=value.items.find(r=>!['completed','failed','cancelled','deleted'].includes(r.status))
    if(pending){const result=await request('operation',{scope:dataScope.value,id:pending.id});Object.assign(pending,result)}
    operationNote.textContent=value.items.slice(0,4).map(r=>r.document.slice(-12)+': '+r.status).join('\n')
  }
  const refresh=async()=>{
    if(!config.endpoint)return
    const value=await request('documents',{scope:dataScope.value,offset});rows=value.items;total=value.total;list.replaceChildren();contents.value=''
    for(const row of rows){const option=make('option',row.id+' · '+row.memory_unit_count+' facts');option.value=row.id;list.append(option)}
    list.selectedIndex=-1;pageNote.textContent=`Showing ${rows.length?offset+1:0}–${offset+rows.length} of ${total} documents.`;await operations()
  }
  button(connectionActions,'Check connection',()=>run(async()=>{token=null;const result=await request('check',{...Object.fromEntries(Object.entries(fields).map(([k,e])=>[k,e.value])),activeScope:scope.value});token=result.token;note.textContent='Connection verified. Save to enable memory.'}))
  const save=button(connectionActions,'Save and enable',()=>run(async()=>{await request('configure',{token});await loadConfig();offset=0;await refresh()}))
  button(connectionActions,'Disable memory',()=>run(async()=>{await request('disable');await loadConfig()}))
  connection.append(make('p','Disable stops new recall and retention. Submitted operations may finish. Retained data stays available for view, export and deletion.'))
  const retain=button(dataActions,'Retain this text',()=>run(async()=>{
    if(!text.value.trim())throw Error('Enter the text to retain.')
    const value=await request('retain',{scope:dataScope.value,content:text.value,provenance:provenance(),requestId:crypto.randomUUID()})
    text.value='';note.textContent=value.status==='pending'?'Retention queued. Refresh to check completion.':'Retention outcome is unknown. Refresh its status before saving again.';await operations()
  }))
  button(dataActions,'Refresh',()=>run(refresh))
  button(dataActions,'Previous',()=>run(async()=>{offset=Math.max(0,offset-20);await refresh()}))
  button(dataActions,'Next',()=>run(async()=>{if(offset+rows.length<total){offset+=20;await refresh()}}))
  list.onchange=()=>run(async()=>{const value=await request('document',{scope:dataScope.value,id:list.value});contents.value=value.original_text??'(No source text)'})
  dataScope.onchange=()=>run(async()=>{offset=0;await refresh()})
  button(dataActions,'Delete selected',()=>run(async()=>{
    if(!list.value)return
    if(!doc.defaultView.confirm('Delete this source document and its associated memories from Hindsight?'))return
    await request('delete',{scope:dataScope.value,id:list.value});await refresh();note.textContent='Memory document deleted.'
  }))
  button(dataActions,'Export facts',()=>run(async()=>{
    const facts=[];let next=0
    while(true){const page=await request('exportPage',{scope:dataScope.value,offset:next});facts.push(...page.items);next+=page.items.length;if(next>=page.total)break;if(!page.items.length)throw Error('Memory changed during export. Please try again.')}
    const url=URL.createObjectURL(new Blob([JSON.stringify({provider:'hindsight',scope:dataScope.value,facts},null,2)],{type:'application/json'}));const a=make('a');a.href=url;a.download='augmentor-memory.json';a.click();setTimeout(()=>URL.revokeObjectURL(url),30000);note.textContent='Memory facts exported.'
  }))
  button(actions,'Done',()=>dialog.close())
  dialog.addEventListener('close',()=>{closed=true;fields.apiKey.value='';dialog.remove()})
  dialog.addEventListener('cancel',()=>{if(busy)note.textContent='Submitted operations remain available in Memory.'})
  presentSettingsForm(dialog,container);void run(async()=>{await loadConfig();if(!closed)await refresh()});return dialog
}

File: apps/native/augmentor_linux/memory.py
# Copyright © 2026 Manolo Remiddi · SPDX-License-Identifier: LicenseRef-Augmentor-MIT-Resale-1.0
"""Shared memory configuration and explicit retention, independent of the harness."""
import json
import os
from pathlib import Path
import tempfile
import threading
import uuid
from PySide6.QtCore import QTimer, Signal, QSignalBlocker
from PySide6.QtWidgets import (QDialog,QVBoxLayout,QFormLayout,QHBoxLayout,QLabel,
    QLineEdit,QPushButton,QComboBox,QPlainTextEdit,QTabWidget,QWidget,QListWidget,
    QFileDialog,QMessageBox)
from .prompt_client import PromptClient
from .dual_memory import DualMemoryPanel


class MemoryDialog(QDialog):
    completed=Signal(object,object)

    def __init__(self,owner):
        super().__init__(owner);self.owner=owner;self.client=PromptClient();self.token=None;self.busy=False;self.dismissed=False;self.config={};self.documents=[];self.offset=0;self.total=0
        self.completed.connect(lambda callback,value:callback(value) if not self.dismissed else None)
        self.setWindowTitle('Memory');self.resize(600,640)
        layout=QVBoxLayout(self)
        intro=QLabel('Automatic relationship and work memory is shared across Augmentor conversations. Hindsight maintains relationship pages and searchable project memory.');intro.setWordWrap(True);layout.addWidget(intro)
        tabs=QTabWidget();self.tabs=tabs;layout.addWidget(tabs);connection=QWidget();data=QWidget();tabs.addTab(DualMemoryPanel(owner),'Automatic');tabs.addTab(connection,'Manual library connection');tabs.addTab(data,'Manual library memories')
        form=QFormLayout(connection)
        self.endpoint=QLineEdit();self.endpoint.setPlaceholderText('http://127.0.0.1:8888')
        self.key=QLineEdit();self.key.setEchoMode(QLineEdit.EchoMode.Password)
        self.user=QLineEdit();self.project=QLineEdit();self.scope=QComboBox();self.scope.addItem('User','user');self.scope.addItem('Project','project')
        self.fields=[self.endpoint,self.key,self.user,self.project,self.scope]
        for title,field in [('Endpoint',self.endpoint),('API key',self.key),('User bank',self.user),('Project bank (optional)',self.project),('Agent recall scope',self.scope)]:
            field.setAccessibleName(title);form.addRow(title,field)
            (field.currentIndexChanged if field is self.scope else field.textChanged).connect(self.invalidate)
        explanation=QLabel('Requires Hindsight 0.9.2. Checking contacts that service without sending chat history. Saving enables recall. Retain sends only the text you enter in Memories; Hindsight may charge for processing. The key is stored in your private configuration without encryption.');explanation.setWordWrap(True);form.addRow(explanation)
        row=QHBoxLayout();self.check=QPushButton('Check connection');self.save=QPushButton('Save and enable');self.save.setEnabled(False)
        self.check.clicked.connect(self.check_connection);self.save.clicked.connect(self.configure);row.addWidget(self.check);row.addWidget(self.save);form.addRow(row)
        self.disable=QPushButton('Disable memory');self.disable.clicked.connect(lambda:self.run('disable',{},self.loaded));form.addRow(self.disable)
        detail=QLabel('Disable stops new recall and retention. Previously submitted operations may finish. Retained data remains available for viewing, export and deletion.');detail.setWordWrap(True);form.addRow(detail)
        body=QVBoxLayout(data);self.data_scope=QComboBox();self.data_scope.addItem('User','user');self.data_scope.addItem('Project','project');self.data_scope.setAccessibleName('Memory data scope');body.addWidget(self.data_scope)
        self.data_scope.currentIndexChanged.connect(self.change_scope)
        self.text=QPlainTextEdit();self.text.setPlaceholderText('Text to remember. Only this text will be retained.');self.text.setAccessibleName('Memory text');self.text.setMaximumHeight(115);body.addWidget(self.text)
        self.retain=QPushButton('Retain this text');self.retain.clicked.connect(self.remember);body.addWidget(self.retain)
        self.operations=QLabel();self.operations.setWordWrap(True);body.addWidget(self.operations)
        self.list=QListWidget();self.list.setAccessibleName('Retained documents');self.list.currentRowChanged.connect(self.read_document);body.addWidget(self.list)
        self.contents=QPlainTextEdit();self.contents.setReadOnly(True);self.contents.setAccessibleName('Retained document text');body.addWidget(self.contents)
        paging=QHBoxLayout();body.addLayout(paging)
        self.previous=QPushButton('Previous');self.next=QPushButton('Next');self.page=QLabel()
        self.previous.clicked.connect(lambda:self.change_page(-20));self.next.clicked.connect(lambda:self.change_page(20))
        paging.addWidget(self.previous);paging.addWidget(self.page,1);paging.addWidget(self.next)
        row=QHBoxLayout();body.addLayout(row)
        self.data_actions=[]
        for title,callback in [('Refresh',self.refresh),('Export facts',self.export),('Delete selected',self.delete)]:
            button=QPushButton(title);button.clicked.connect(callback);row.addWidget(button);self.data_actions.append(button)
        self.note=QLabel('Loading memory settings…');self.note.setWordWrap(True);layout.addWidget(self.note)
        close=QPushButton('Done');close.clicked.connect(self.reject);layout.addWidget(close)
        self.timer=QTimer(self);self.timer.setInterval(3000);self.timer.timeout.connect(lambda:self.poll_operations() if self.tabs.currentIndex()==2 else None);self.timer.start()
        self.run('describe',{},self.loaded)

    def invalidate(self,*_):self.token=None;self.save.setEnabled(False)

    def background(self,work,finished):
        # The companion is available before a harness connects, including when
        # the window is opened without a controller.
        def execute():
            value=work()
            try:self.completed.emit(finished,value)
            except RuntimeError:pass  # The user closed and destroyed the dialog.
        threading.Thread(target=execute,daemon=True).start()

    def set_busy(self,busy):
        self.busy=busy
        for field in self.fields+[self.data_scope,self.list,self.disable]+self.data_actions:field.setEnabled(not busy)
        self.check.setEnabled(not busy);self.save.setEnabled(not busy and bool(self.token))
        self.retain.setEnabled(not busy and bool(self.config.get('enabled')))
        self.previous.setEnabled(not busy and self.offset>0);self.next.setEnabled(not busy and self.offset+20<self.total)

    def run(self,action,p,callback,request_id=None):
        if self.busy:return
        self.set_busy(True)
        def work():
            try:return self.client.call('memory.'+action,p,request_id=request_id),None
            except Exception as exc:return None,str(exc)
        def finished(value):
            if self.dismissed:return
            self.set_busy(False)
            if value[1]:self.note.setText(value[1])
            else:callback(value[0])
        self.background(work,finished)

    def loaded(self,value):
        self.config=value
        self.endpoint.setText(value.get('endpoint',''));self.user.setText(value.get('userBank','augmentor-user-'+uuid.uuid4().hex[:12]))
        self.project.setText(value.get('projectBank',''));self.scope.setCurrentIndex(self.scope.findData(value.get('activeScope','user')))
        with QSignalBlocker(self.data_scope):self.data_scope.setCurrentIndex(self.data_scope.findData(value.get('activeScope','user')))
        self.offset=0
        self.key.clear();self.key.setPlaceholderText('Stored key kept if left blank' if value.get('apiKeySet') else 'Required for a remote service')
        self.retain.setEnabled(bool(value.get('enabled')));self.note.setText('Hindsight memory enabled.' if value.get('enabled') else 'Hindsight memory disabled.')
        if value.get('endpoint'):self.refresh()

    def check_connection(self):
        self.invalidate();self.note.setText('Checking Hindsight…')
        def checked(value):self.token=value['token'];self.save.setEnabled(True);self.note.setText('Connection verified. Save to enable memory.')
        self.run('check',{'endpoint':self.endpoint.text(),'apiKey':self.key.text(),'userBank':self.user.text(),'projectBank':self.project.text(),'activeScope':self.scope.currentData()},checked)

    def configure(self):
        if self.token:self.run('configure',{'token':self.token},self.loaded)

    def remember(self):
        content=self.text.toPlainText()
        if not content.strip():self.note.setText('Enter the text to retain.');return
        provenance={'surface':'linux','harness':getattr(self.owner.controller,'harness','pi')}
        session=getattr(self.owner.controller,'session',None)
        if session:provenance['sessionId']=session
        def saved(value):
            self.text.clear();self.note.setText('Retention queued. Refresh to check completion.' if value['status']=='pending' else 'Retention outcome is unknown. Refresh its status before saving again.');self.poll_operations()
        self.run('retain',{'scope':self.data_scope.currentData(),'content':content,'provenance':provenance},saved,uuid.uuid4().hex)

    def refresh(self,*_):
        if not self.config.get('endpoint') or self.busy:return
        def loaded(value):
            self.documents=value['items'];self.total=value['total'];self.list.clear();self.contents.clear()
            for row in self.documents:self.list.addItem(str(row['id'])+' · '+str(row.get('memory_unit_count',0))+' facts')
            self.page.setText(f"{self.offset+1 if self.documents else 0}–{self.offset+len(self.documents)} of {self.total}")
            self.set_busy(False)
            self.note.setText('Export facts includes every page.')
        self.run('documents',{'scope':self.data_scope.currentData(),'offset':self.offset},loaded)

    def change_scope(self,*_):
        self.offset=0;self.documents=[];self.list.clear();self.contents.clear();self.refresh()

    def change_page(self,delta):
        if self.busy:return
        self.offset=max(0,self.offset+delta);self.refresh()

    def read_document(self,index):
        if index<0 or index>=len(self.documents):return
        self.run('document',{'scope':self.data_scope.currentData(),'id':self.documents[index]['id']},lambda r:self.contents.setPlainText(r.get('original_text') or '(No source text)'))

    def poll_operations(self):
        if self.busy or not self.config.get('endpoint'):return
        scope=self.data_scope.currentData()
        def receive(value):
            self.operations.setText('\n'.join(row['document'][-12:]+': '+row['status'] for row in value['items'][:4]))
            pending=next((r for r in value['items'] if r['status'] not in ('completed','failed','cancelled','deleted')),None)
            if pending:self.run('operation',{'scope':scope,'id':pending['id']},lambda r:self.operations.setText(r['document'][-12:]+': '+r['status']))
        self.run('operations',{'scope':scope},receive)

    def delete(self):
        index=self.list.currentRow()
        if self.busy or not 0<=index<len(self.documents):return
        row=self.documents[index]
        if QMessageBox.question(self,'Delete memory','Delete this source document and its associated memories from Hindsight?')!=QMessageBox.StandardButton.Yes:return
        self.run('delete',{'scope':self.data_scope.currentData(),'id':row['id']},lambda _:self.refresh())

    def export(self):
        if self.busy:return
        path,_=QFileDialog.getSaveFileName(self,'Export memory facts','augmentor-memory.json','JSON (*.json)')
        if not path:return
        scope=self.data_scope.currentData();self.set_busy(True);self.note.setText('Exporting facts…')
        def work():
            temporary=None
            try:
                rows=[];offset=0
                while True:
                    page=self.client.call('memory.exportPage',{'scope':scope,'offset':offset});items=page['items'];rows.extend(items);offset+=len(items)
                    if offset>=page['total']:break
                    if not items:raise RuntimeError('Memory changed during export. Please try again.')
                descriptor,temporary=tempfile.mkstemp(prefix='.augmentor-memory-',dir=Path(path).parent)
                with os.fdopen(descriptor,'w') as file:
                    json.dump({'provider':'hindsight','scope':scope,'facts':rows},file,ensure_ascii=False,indent=2)
                    file.flush();os.fsync(file.fileno())
                os.replace(temporary,path);temporary=None
                return None
            except Exception as exc:return str(exc)
            finally:
                if temporary:Path(temporary).unlink(missing_ok=True)
        def finished(error):
            if self.dismissed:return
            self.set_busy(False);self.note.setText(error or 'Memory facts exported.')
        self.background(work,finished)

    def reject(self):
        self.dismissed=True;self.timer.stop();self.key.clear();super().reject()
